formats/properties/commonMain/src/kotlinx/serialization/Properties.kt

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.CompositeDecoder.Companion.READ_DONE
+import kotlinx.serialization.modules.*
+
+
+/**
+ * Transforms a [Serializable] class' properties to a single flat [Map] which consists of
+ * string keys and primitive type values, and vice versa.
+ *
+ * If the given class has non-primitive property `d` of arbitrary type `D`, `D` values are inserted
+ * into the same map; keys for such values are prefixed with string `d.`:
+ *
+ * ```kotlin
+ * @Serializable
+ * class Data(val property1: String)
+ *
+ * @Serializable
+ * class Recursive(val data: Data, val property2: String)
+ *
+ * val map = Mapper.store(Recursive(Data("value1"), "value2"))
+ * // map contents will be the following:
+ * // property2 = value2
+ * // data.property1 = value1
+ * ```
+ *
+ * If the given class has a [List] property `l`, each value from the list
+ * would be prefixed with `l.N.`, where N is an index for particular value.
+ * Additional `l.size` property with a list size would be added.
+ * [Map] is treated as a [key,value,...] list.
```

Comment:
From the user perspective, why one would want to see `l.size` in the output?
Let's omit this information and use `-1` as size instead?

Reply:
Do you mind if I leave this to a follow-up PR as a feature request?

Reply:
Sure